infra/src/main/kotlin/fi/fta/geoviite/infra/dataImport/SwitchLibraryDataMigration.kt

```diff
@@ -158,3 +158,25 @@ class V10_03_06__SwitchLibraryDataMigration : BaseJavaMigration() {
     // Increase this manually when switch library data changes
     override fun getChecksum(): Int = 6
 }
+
+@Suppress("unused", "ClassName")
+class V53__add_rr54_4x_switch_structure : BaseJavaMigration() {
+    val logger: Logger = LoggerFactory.getLogger(this::class.java)
+
+    override fun migrate(context: Context?) {
+        withUser(ImportUser.SWITCH_LIB_IMPORT) {
+            logger.info("Add RR54-4x1:9 structure into library")
+
+            val connection =
+                context?.connection ?: throw IllegalStateException("Can't run imports without DB connection")
+            val jdbcTemplate = NamedParameterJdbcTemplate(
+                SingleConnectionDataSource(connection, true)
+            )
+            val switchStructureDao = SwitchStructureDao(jdbcTemplate)
+            switchStructureDao.insertSwitchStructure(RR54_4x1_9())
```

Comment:
Olisikohan tälle vaikeaa tehdä upsert-operaatio dao:on? Tämän versioidun migraation voisi nimittäin vaihtaa repeatableksi jos tuo insert olisi upsert (idempotentti). Sitten jos tuo data muuttuu niin ihan oikeasti olisi mahdollista  vain nostaa tuota checksummia ja data päivittyy.

Reply:
Tuo structure taitaa kyllä olla moniosainen, joten tämä ei välttämättä ole ihan triviaalia. Jos menee liian vaikeaksi niin ei tarvitse tähän fiksiin sitä tehdä, mutta se voisi olla pitkässä juoksussa parempi malli kuin tehdä nämä structurepäivitykset versioituina. Ainoa ongelma minkä keksin on että jos joku versioitu migraatio vaatii jonkun tietyn structuren niin sitten se täytyy olla aiemmassa versioidussa tehty (repeatablet tulee vasta migraatioiden lopuksi).

Reply:
Vaihdetyyppikorjauksissa taitaa olla aina vähän tuli pepun alla, joten veikkaan että vaikka nämä onkin kategorisesti huonoja hetkiä tuontyyliselle jumpalle, niin tuskin sitä parempiakaan kauheasti tulee tulevaisuudessa. Sinällään upsert ei kuulosta _niin_ monimutkaiselta rakenteelta (monimutkaisin osuus lienee jointtien tunnistaminen ja tuolla reviewin pohjallahan oli jo update jossa niin tehdään)